src/lib/mongoBackend/MongoCommonUpdate.cpp

```diff
@@ -254,287 +241,178 @@ static bool equalMetadataVectors(BSONObj& mdV1, BSONObj& mdV2) {
-         * in database by a CB instance previous to the support of creation and modification dates */
-        if (attr.hasField(ENT_ATTRS_CREATION_DATE)) {
-            newAttr->append(ENT_ATTRS_CREATION_DATE, attr.getIntField(ENT_ATTRS_CREATION_DATE));
-        }
-        if (attr.hasField(ENT_ATTRS_MODIFICATION_DATE)) {
-            newAttr->append(ENT_ATTRS_MODIFICATION_DATE, attr.getIntField(ENT_ATTRS_MODIFICATION_DATE));
-        }
-        string id = STR_FIELD(attr, ENT_ATTRS_ID);
-        if (id != "") {
-            newAttr->append(ENT_ATTRS_ID, STR_FIELD(attr, ENT_ATTRS_ID));
-        }
-        deleted = false;
+      ab.append(ENT_ATTRS_MODIFICATION_DATE, attr.getIntField(ENT_ATTRS_MODIFICATION_DATE));
     }
+  }
 
-    return deleted;
+  *mergedAttr = ab.obj();
+  return actualUpdate;
 }
 
 /* ****************************************************************************
 *
 * updateAttribute -
 *
-* Returns true if an attribute was found and replaced, false otherwise. If true,
+* Returns true if an attribute was found, false otherwise. If true,
 * the "actualUpdate" argument (passed by reference) is set to true in the case that the
 * original value of the attribute was different that the one used in the update (this is
 * important for ONCHANGE notifications)
 *
 */
-static bool updateAttribute(BSONObj& attrs, BSONObj& newAttrs, ContextAttribute* caP, bool& actualUpdate) {
-
-    BSONArrayBuilder newAttrsBuilder;
-    actualUpdate = false;
-    bool updated = false;
-    for( BSONObj::iterator i = attrs.begin(); i.more(); ) {
+static bool updateAttribute(BSONObj& attrs, BSONObjBuilder* toSet, ContextAttribute* caP, bool& actualUpdate)
+{
+  actualUpdate = false;
 
-        BSONObjBuilder newAttr;
-        bool unitActualUpdate = false;
-        if (checkAndUpdate(newAttr, i.next().embeddedObject(), *caP, &unitActualUpdate) && !updated) {
-            updated = true;
-        }
-        /* If at least one actual update was done at checkAndUpdate() level, then updateAttribute()
-         * actual update is true */
-        if (unitActualUpdate == true) {
-            actualUpdate = true;
-        }
+  /* Attributes with metadata ID are stored as <attrName>_<ID> in the attributes embedded document */
+  std::string effectiveName = caP->name;
+  if (caP->getId() != "")
+  {
+    effectiveName += "__" + caP->getId();
+  }
 
-        newAttrsBuilder.append(newAttr.obj());
+  //LM_T(LmtMongo, ("==== %s", attrs.toString().c_str()));
+  if (attrs.hasField(effectiveName.c_str()))
+  {
+    BSONObj attr = attrs.getField(effectiveName).embeddedObject();
+    BSONObj newAttr;
+    actualUpdate = mergeAttrInfo(attr, caP, &newAttr);
+    if (actualUpdate)
+    {
+      const std::string composedName = std::string(ENT_ATTRS) + "." + effectiveName;
+      toSet->append(composedName, newAttr);
     }
-    newAttrs = newAttrsBuilder.arr();
-
-    return updated;
+    return true;
+  }
+  else
+  {
+    return false;
+  }
 
```

Comment:
These four lines can be replaced by one line.

Even better would be to check for the condition and return false if not fulfilled, before starting (gaining indents):

    if (!attrs.hasField(effectiveName.c_str()))
    {
      return false:
    }
    
    BSONObj attr = attrs.getField(effectiveName).embeddedObject();
    ...